keps/sig-scheduling/4671-gang-scheduling/README.md

```diff
@@ -469,19 +519,269 @@ states (e.g. not yet block resources) will help with improving the scheduling ac
+or strictly at the `PodGroup` level.
+
+* PodGroup Level: The cycle processes only the specific `PodGroup` (and replica key)
+  associated with the popped Pod. This is simpler and aligns with
+  the Gang Scheduling definition and current implementation.
+* Workload Level: The cycle attempts to schedule all PodGroups within the Workload.
+  This allows for complex dependencies between groups but increases the complexity
+  and doesn't bring immediate added value.
+
+*Proposed:* Implement it on PodGroup Level for Beta. However, future migration
+to the Workload Level might necessitate non-trivial changes to the phase
+introduced by this KEP.
+<<[/UNRESOLVED]>>
+```
+
+The cycle proceeds as follows:
+
+1. The scheduler takes either pod group itself or its Pod representative from
+   the scheduling queue. If the pod group is unscheduled, it temporarily removes
+   all group's pods from the queue for processing. The order of processing
+   is determined by the queueing mechanism (see *Queuing and Ordering* below).
+   
+2. A single cluster state snapshot is taken for the entire group operation
+   to ensure consistency during the cycle.
+
+3. The scheduler runs a specialized algorithm (detailed below)
+   to find placements for the group.
+
+4. Outcome:
+   * If the group (i.e., at least `minCount` Pods) can be placed,
+     these Pods have the `.status.nominatedNodeName` set.
+     They are then effectively "reserved" on those nodes in the
+     scheduler's internal cache. Pods are then pushed to the
+     active queue (restoring their original timestamps to ensure fairness)
+     to pass through the standard scheduling and binding cycle,
+     which will respect the nomination.
+   * If `minCount` cannot be met (even after calculating potential
+     preemptions), the scheduler rejects the entire group. Standard backoff
+     logic applies (see *Failure Handling*), and Pods are returned to
+     the scheduling queue.
+
+#### Queuing and Ordering
+
+Workload-aware preemption (an `Alpha` effort in [KEP-5710](https://github.com/kubernetes/enhancements/pull/5711))
+will introduce a specific scheduling priority for a workload.
+Having that in mind, it is beneficial to design a queueing mechanism open
+for taking a workload's scheduling priority into account.
+However, as we need to support ordering before that feature can be enabled,
+we also need to derive the priority from the pod group's pods.
+One such formula can be to set it to the highest priority found within the pod group.
+
+```md
+<<[UNRESOLVED Queue Implementation Strategy]>>
+To ensure that we process the pod group (replica) at an appropriate time and
+don't starve other pods (including gang pods in the pod-by-pod scheduling phase)
+from being scheduled, we need to have a good queueing mechanism for pod groups.
+There are several alternatives:
+
+Alternative 1 (Modify sorting logic):
+
+Modify the sorting logic within the existing `PriorityQueue` to put all pods
```

Comment:
We would need to modify it once we have workload priority, right? Otherwise we can keep the current sorting algorithm and pull all gang members from the queue once we encounter the first member of the gang (and switch to workload scheduling), right?